src/main/scala/net/snowflake/spark/snowflake/LoggerWrapper.scala

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2015-2020 Snowflake Computing
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.snowflake.spark.snowflake
```

Comment:
```
class TelemetryReporter(val telemetryEnabled) {
  def sendLogTelemetry() = {
    ...
  }
}
```

```
class LoggerWrapper(logger: Logger, telemetryReporter: telemetryReporter) {
}
```

And you don't need loggerwrapperfactory or anything like that.

I also don't understand how you will send in-band telemetry, don't you need a connection or a session token for that ? How will this design work in that case ?

Reply:
1. LoggerWrapperFactory is optional, it is introduced to align the LoggerFactory pattern.

2. We need a capability to turn on/off to send telemetry message. So we need "object LoggerWrapper {}" in my case. If we want to use "TelemetryReporter", we may need "object TelemetryReporter".

3. In general, most of the classes work in spark driver side, only a few classes work in executor side. LoggerWrapper by default will regarded log as driver log. If the logger is on spark executor, the executor can call a new method e.g. LoggerWrapper.setExecutorMetadata(meta) to the logger wrapper. The "meta" will include executor_id, partition_id, session_token, etc.

4. In general, introducing TelemetryReporter is a good idea. It can decouple the logger wrapper and telemetry enable/disable/sending

Reply:
As a summary, I will update to use TelemetryReporter and LoggerWrapper to decouple the logging and telemetry. In the same time, I prefer keeping LoggerWrapperFactory for better alignment with Logger factory pattern.

Reply:
If you make `TelemetryReporter` serializable, you can just send the session token over to the executors by serializing the reporter object.

Reply:
Good idea. We can do it later.